### ISSUE TITLE
migrate from travis to github actions

### ISSUE THREAD

**Issue by hohwille** [devonfw/devon4j#347]
For several reasons we have finally decided to migrate from travis-ci to github actions.

**Comment by hohwille** [comment]
I just reviewed PR #343 and was confused about the secrets. Now I learned that we can not have secrets for PR builds. This will prevent us from implementing PR #294 so I am not convinced that we should go away from travis if we can not solve this with github actions...

**Comment by hohwille** [comment]
OK, it seems possible after all:
* https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target
* https://github.com/SonarSource/sonarcloud-github-action

So let us try this out...

**Comment by hohwille** [comment]
@ssarmokadam can you please asap get in touch with @sujith-mn 
He has created the github actions for devonfw-ide:
https://github.com/devonfw/ide/pull/529/files

We do not seem to be aligned here and follow different approaches and conventions.

**Comment by maybeec** [comment]
@ssarmokadam , @sujith-mn , please check how it is done in solicitor as we discussed it might be the most complete and working setup. Please make sure the files apply the same standards and names. Even I just made quite some reviews and sometimes there is the email coming from the secrets, sometimes not. It should be same for all. Please make sure, we have the same view on it and have a continuous chat on that do not do the work twice ;)

**Comment by hohwille** [comment]
Was autoclosed by PR. Will reopen as not yet done.
Only after all working smoth travis disabled and  `.travis.yml` being deleted, we will close this issue.

**Comment by hohwille** [comment]
@ssarmokadam have you talked with @sujith-mn about this?
I have completed github actions in devonfw-ide. You can get the changes from there to complete this in devon4j.
https://github.com/devonfw/ide/pull/533

**Comment by hohwille** [comment]
Reopened due to the review comment in #360 that suggests a final improvement.

**Comment by hohwille** [comment]
We still have the old github action in place: `deploy.yml` was not renamed to `build.yml` but instead has been copied.
So we need to delete `deploy.yml`...

**Comment by hohwille** [comment]
I just fixed the badge that has been forgotten to change and still was pointing to travis.
However, still this does not seem to work as it also runs in PRs:
https://github.com/devonfw/devon4j/actions/runs/675514421/workflow#L26

**Comment by ssarmokadam** [comment]
@hohwille It looks like access issue ... SONATYPE_USERNAME: ${{ secrets.SONATYPE_USERNAME }}
          SONATYPE_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }} not getting configured properly

**Comment by hohwille** [comment]
> It looks like access issue ... SONATYPE_USERNAME: ${{ secrets.SONATYPE_USERNAME }}
SONATYPE_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }} not getting configured properly

Yes, any ideas what could be wrong. It is really hard to debug this. Maven does not even log if USERNAME or PASSWORD is configured at all. But 403 indicates to me that authentication worked but something failed with the permissions.
Therefore I think that the values of these secrets are actually wrong and that account cannot deploy `com.devonfw` to OSSRH.

**Comment by hohwille** [comment]
@maybeec was able to fix the 403 problem with OSSRH deployment. 👍

**Comment by hohwille** [comment]
However, the condition to run a step if not a PR is still not working.
In travis there was a reliable documentation available but on github the documentation of actions seems wrong or missleading.
This is also not working at all:

https://github.com/devonfw/devon4j/blob/7232b7605261a291be3dd67bbdf0f03f96109438/.github/workflows/build.yml#L26

Maybe something is still wrong with the YAML syntax and the step is executed independent of the if clause. Any hint would be highly appreciated. I start to seriously hate these github actions...

**Comment by hohwille** [comment]
Seems I finally found it:
https://tjtelan.com/blog/github-actions-push-vs-pr-workflow/

**Comment by hohwille** [comment]
GHA sucks:
https://github.community/t/run-step-on-if-branch-tag-is/16965

**Comment by hohwille** [comment]
https://www.edwardthomson.com/blog/github_actions_13_conditionals.html

**Comment by hohwille** [comment]
My last try was this:
https://github.com/devonfw/devon4j/blob/a69b68b9f4c2af94d52da3ddcc4b34240088b6d3/.github/workflows/build.yml#L26

Still not working and running on PR. Giving up...

**Comment by hohwille** [comment]
@maybeec fixed this with a very nice solution:
* we have `build.yml` for changes on master branch that also deploys to nexus, etc.
* we have `build-pr.yml` for the build of PRs

This causes a minimum redundancy of build steps but it is a perfect KISS solution and is easy to understand and maintain.
Thanks @maybeec for this 👍

**Comment by hohwille** [comment]
Github actions are a never ending nightmare.
They still do not work for PRs as they are building the code without the changes from the PR:
https://github.com/devonfw/devon4j/pull/339
I have created a fresh PR with that test added to retry with the same effect:
https://github.com/devonfw/devon4j/pull/376

I assume this `on: pull_request_target` feature is simply broken and does not work.

**Comment by hohwille** [comment]
I had to remove sonarcloud analysis for PRs. It is simply impossible to solve with GHA. Sucks!

**Comment by maybeec** [comment]
Just for your understanding https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

**Comment by hohwille** [comment]
So we do not get static code analysis and sonar quality checks for PRs. If the risk of loosing the sonarcloud password and letting some hacker publish false analysis results is IMHO rediculous - we can at any time simply reset the account and create a new token while the hacker has absolutely no benefit from spending the effort except wasting some of his and our time. So I am closing. If someone sees some opportunity to get the sonar analysis for PRs, we can create a new issue and proceed from there.

**Comment by maybeec** [comment]
Related to
https://community.sonarsource.com/t/sonarcloud-github-action-for-a-monorepo-and-support-for-forked-prs/37157/12
And
https://jira.sonarsource.com/browse/MMF-1371
